neuralhydrology/modelzoo/transformer.py

```diff
@@ -0,0 +1,167 @@
+    -- seq_length : integer number of timesteps to treat in the input sequence.
+    -- transformer_nheads : number of self-attention heads.
+    -- transformer_dim_feedforward : dimension of the feed-fowrard networks between self-attention heads.
+    -- transformer_dropout: dropout in the feedforward networks between self-attention heads.
+    -- transformer_nlayers: number of stacked self-attention + feedforward layers.
+
+    Parameters
+    ----------
+    cfg : Config
+        The run configuration.
+    """
+
+    def __init__(self, cfg: Dict):
+        super(Transformer, self).__init__(cfg=cfg)
+
+        if cfg.embedding_hiddens:
+            LOGGER.warning("## Warning: Embedding settings are ignored. Use EmbCudaLSTM for embeddings")
+
+        input_size = len(cfg.dynamic_inputs + cfg.evolving_attributes + cfg.hydroatlas_attributes +
+                         cfg.static_attributes)
+        if cfg.use_basin_id_encoding:
+            input_size += cfg.number_of_basins
+
+        if cfg.head.lower() == "umal":
+            input_size += 1
+
+        # embedding 
+        self.embedding_dim = cfg.transformer_embedding_dimension
+        self.embedding = nn.Linear(in_features=input_size, 
+                                   out_features=self.embedding_dim)
+
+        # positional encoder
+        self.posistional_encoding_type = cfg.transformer_positional_encoding_type
+        if self.posistional_encoding_type.lower() in ['concat', 'cat', 'concatenate']:
+            self.encoder_dim = self.embedding_dim*2
+        elif self.posistional_encoding_type.lower() in ['sum', 'add', 'addition']:
+            self.encoder_dim = self.embedding_dim
+        else:
+            raise RuntimeError(f"Unrecognized positional encoding type: {self.posistional_encoding_type}")
+        self.pos_encoder = PositionalEncoding(d_model=self.embedding_dim, 
+                                              dropout=cfg.transformer_positional_dropout, 
+                                              max_len=cfg.seq_length)
+
+        # positional mask
+        self.mask = None
+
+        # encoder
+        encoder_layers = nn.TransformerEncoderLayer(d_model=self.encoder_dim, 
+                                                    nhead=cfg.transformer_nheads, 
+                                                    dim_feedforward=cfg.transformer_dim_feedforward, 
+                                                    dropout=cfg.transformer_dropout)
+        self.encoder = nn.TransformerEncoder(encoder_layer=encoder_layers, 
+                                             num_layers=cfg.transformer_nlayers,
+                                             norm=None) 
+
+        # head (instead of a decoder)
+        self.dropout = nn.Dropout(p=cfg.output_dropout)
+        self.head = get_head(cfg=cfg, n_in=self.encoder_dim, n_out=self.output_size)
+
+        # init weights and biases 
+        self._reset_parameters()
+
+    def _reset_parameters(self): 
+        initrange = 0.1
```

Comment:
Out of curiosity: Is this anything you tested yourself? Are the PyTorch defaults not good? Is there any reference for that? I could see why this hard-coded initialization might raise questions in a few month from now, when no one remembers why it is there. So if there is a reason for that (I am sure there is) you could eventually add a one-line comment.

Reply:
@grey-nearing Can you clarify on this?